### PR TITLE
nixos/syncthing: match stale ids by subtracting in jq

### LINKED ISSUE #259256: syncthing fails to delete old folders
### Describe the bug
if a folder is a substring of another folder, we don't delete it even if we have overrideFolders and the substring id got removed from the config.

### Steps To Reproduce
1. define a folder with id "xxx"
2. define a folder with id "yyyzzzxxx";
3. remove folder with id "xxx" from config
4. it never gets removed

### Expected behavior
folder "xxx" should have been removed after removing it from config 

### Notify maintainers
@doronbehar 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### PR DESCRIPTION
**Pull request by Lassulus**
## Description of changes

https://github.com/NixOS/nixpkgs/issues/259256

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Lassulus** [comment]
another problem which is not fixed here: it's not possible to delete folders with a / in the ID

**Comment by Lassulus** [comment]
hmm, it seems like this approach is flawed. In the frontend the PUT endpoint is used, since that one is capable of escaping the IDs correctly. I guess we should do the same?

File: nixos/modules/services/networking/syncthing.nix [hunk 1]
```diff
@@ -51,79 +55,82 @@ let
...
-        # only if s.override == true then we DELETE the relevant folders
-        # afterwards.
-        (map (new_cfg: ''
-          curl -d ${lib.escapeShellArg (builtins.toJSON new_cfg)} -X POST ${s.baseAddress}
-        ''))
+
+    ${if (cfg.overrideDevices && cfg.overrideFolders) then ''
+      # If we need to override both devices and folders, we can just use a single PUT request to push the new config
+      curl -X PUT ${cfg.guiAddress}/rest/config -d ${lib.escapeShellArg (builtins.toJSON newConfig)}
+    '' else
```
Comment (doronbehar) [inline nixos/modules/services/networking/syncthing.nix:62]: I don't like how the behavior of this script changes fundamentally when someone changes these `override{Folder,Device}s`...

**Review by doronbehar — COMMENTED** [review]
Although the diff is large, with `git show -w --word-diff-regex='[^[:space:]]'` it isn't that big, and I don't see how it solves #259256 in case the user has only one of the `override{Folder,Device}s` settings set.

Perhaps you could write a test that proves the issue?

File: nixos/modules/services/networking/syncthing.nix [hunk 2]
```diff
@@ -100,13 +100,13 @@ let
...
-          else
-            curl -X DELETE ${s.baseAddress}/$id
-          fi
+        stale_${conf_type}_ids="$(curl -X GET ${s.baseAddress} | ${jq} \
+          --argjson new_ids ${lib.escapeShellArg (builtins.toJSON s.new_conf_IDs)} \
+          --raw-output \
+          '[.[].${s.GET_IdAttrName}] - $new_ids | .[]'
+        )"
+        for id in ''${stale_${conf_type}_ids}; do
+          curl -X DELETE ${s.baseAddress}/$id
```
Comment (doronbehar) [inline nixos/modules/services/networking/syncthing.nix]: Oh when I ran `git show` previously I didn't notice this change was done in a commit of it's own. This one looks very good. I wonder why it isn't enough to fix the issue.

**Comment by doronbehar** [comment]
I removed the 2nd commit from this PR. And marked this PR to fix #259256. Will open a new issue for the `curl -X DELETE` topic downstream and upstream today.

**Comment by doronbehar** [comment]
@ofborg test syncthing

@ofborg test syncthing-init

@ofborg test syncthing-no-settings

**Pull request merged**